File: apps/sdkwork-iam-flutter-mobile/packages/sdkwork_iam_flutter_mobile_core/lib/src/auth_runtime_metadata.dart
enum SdkworkAuthLoginMethod {
  emailCode,
  password,
  phoneCode,
  sessionBridge,
}

enum SdkworkAuthRegisterMethod { email, phone }

enum SdkworkAuthRecoveryMethod { email, phone }

enum SdkworkAuthOAuthProviderRegion { mainland, overseas }

class SdkworkAuthVerificationPolicyConfig {
  const SdkworkAuthVerificationPolicyConfig({
    this.emailCodeLoginEnabled,
    this.emailRegistrationVerificationRequired,
    this.phoneCodeLoginEnabled,
    this.phoneRegistrationVerificationRequired,
    this.oauthLoginEnabled,
  });

  final bool? emailCodeLoginEnabled;
  final bool? emailRegistrationVerificationRequired;
  final bool? phoneCodeLoginEnabled;
  final bool? phoneRegistrationVerificationRequired;
  final bool? oauthLoginEnabled;
}

class SdkworkAuthRuntimeConfig {
  const SdkworkAuthRuntimeConfig({
    this.loginMethods = const [],
    this.oauthLoginEnabled = false,
    this.oauthProviderRegion,
    this.oauthProviders = const [],
    this.sdkworkOAuthProviderEnabled,
    this.qrLoginEnabled,
    this.recoveryMethods = const [],
    this.registerMethods = const [],
    this.verificationPolicy,
  });

  final List<SdkworkAuthLoginMethod> loginMethods;
  final bool oauthLoginEnabled;
  final SdkworkAuthOAuthProviderRegion? oauthProviderRegion;
  final List<String> oauthProviders;
  final bool? sdkworkOAuthProviderEnabled;
  final bool? qrLoginEnabled;
  final List<SdkworkAuthRecoveryMethod> recoveryMethods;
  final List<SdkworkAuthRegisterMethod> registerMethods;
  final SdkworkAuthVerificationPolicyConfig? verificationPolicy;
}

bool isSdkworkAuthLoginMethod(String? value) {
  return value == 'password'
      || value == 'phoneCode'
      || value == 'emailCode'
      || value == 'sessionBridge';
}

bool isSdkworkAuthRegisterMethod(String? value) {
  return value == 'email' || value == 'phone';
}

bool isSdkworkAuthRecoveryMethod(String? value) {
  return value == 'email' || value == 'phone';
}

bool isSdkworkAuthOAuthProviderRegion(String? value) {
  return value == 'mainland' || value == 'overseas';
}

SdkworkAuthLoginMethod? parseSdkworkAuthLoginMethod(String? value) {
  switch (value) {
    case 'password':
      return SdkworkAuthLoginMethod.password;
    case 'phoneCode':
      return SdkworkAuthLoginMethod.phoneCode;
    case 'emailCode':
      return SdkworkAuthLoginMethod.emailCode;
    case 'sessionBridge':
      return SdkworkAuthLoginMethod.sessionBridge;
    default:
      return null;
  }
}

SdkworkAuthRegisterMethod? parseSdkworkAuthRegisterMethod(String? value) {
  switch (value) {
    case 'email':
      return SdkworkAuthRegisterMethod.email;
    case 'phone':
      return SdkworkAuthRegisterMethod.phone;
    default:
      return null;
  }
}

SdkworkAuthRecoveryMethod? parseSdkworkAuthRecoveryMethod(String? value) {
  switch (value) {
    case 'email':
      return SdkworkAuthRecoveryMethod.email;
    case 'phone':
      return SdkworkAuthRecoveryMethod.phone;
    default:
      return null;
  }
}

SdkworkAuthOAuthProviderRegion? parseSdkworkAuthOAuthProviderRegion(String? value) {
  switch (value) {
    case 'mainland':
      return SdkworkAuthOAuthProviderRegion.mainland;
    case 'overseas':
      return SdkworkAuthOAuthProviderRegion.overseas;
    default:
      return null;
  }
}

SdkworkAuthRuntimeConfig resolveSdkworkAuthRuntimeConfigFromMetadata(
  Map<String, dynamic>? authConfig,
) {
  final verificationPolicy = _resolveMetadataVerificationPolicy(authConfig);
  final configuredLoginMethods = _readStringList(authConfig, 'loginMethods');
  final loginMethods = configuredLoginMethods.isNotEmpty
      ? configuredLoginMethods
          .where(isSdkworkAuthLoginMethod)
          .map(parseSdkworkAuthLoginMethod)
          .whereType<SdkworkAuthLoginMethod>()
          .where((method) {
            if (method == SdkworkAuthLoginMethod.emailCode
                && verificationPolicy.emailCodeLoginEnabled == false) {
              return false;
            }
            if (method == SdkworkAuthLoginMethod.phoneCode
                && verificationPolicy.phoneCodeLoginEnabled == false) {
              return false;
            }
            return true;
          })
          .toList(growable: false)
      : _deriveLoginMethods(authConfig, verificationPolicy);

  final configuredRegisterMethods = _readStringList(authConfig, 'registerMethods')
      .where(isSdkworkAuthRegisterMethod)
      .map(parseSdkworkAuthRegisterMethod)
      .whereType<SdkworkAuthRegisterMethod>()
      .toList(growable: false);
  final supportsLocalCredentials = _readBool(authConfig, 'supportsLocalCredentials') ?? true;
  final registerMethods = configuredRegisterMethods.isNotEmpty
      ? configuredRegisterMethods
      : (supportsLocalCredentials
          ? const [SdkworkAuthRegisterMethod.email]
          : const <SdkworkAuthRegisterMethod>[]);

  final recoveryMethods = _readStringList(authConfig, 'recoveryMethods')
      .where(isSdkworkAuthRecoveryMethod)
      .map(parseSdkworkAuthRecoveryMethod)
      .whereType<SdkworkAuthRecoveryMethod>()
      .toList(growable: false);

  final oauthProviderRegionValue = authConfig?['oauthProviderRegion'] as String?;
  final oauthProviderRegion = parseSdkworkAuthOAuthProviderRegion(oauthProviderRegionValue);

  return SdkworkAuthRuntimeConfig(
    loginMethods: loginMethods,
    oauthLoginEnabled: _readBool(authConfig, 'oauthLoginEnabled') ?? false,
    oauthProviderRegion: oauthProviderRegion,
    oauthProviders: List<String>.from(_readStringList(authConfig, 'oauthProviders')),
    sdkworkOAuthProviderEnabled: _readBool(authConfig, 'sdkworkOAuthProviderEnabled'),
    qrLoginEnabled: _readBool(authConfig, 'qrLoginEnabled'),
    recoveryMethods: recoveryMethods,
    registerMethods: registerMethods,
    verificationPolicy: _hasVerificationPolicyValues(verificationPolicy)
        ? verificationPolicy
        : null,
  );
}

List<SdkworkAuthLoginMethod> _deriveLoginMethods(
  Map<String, dynamic>? authConfig,
  SdkworkAuthVerificationPolicyConfig verificationPolicy,
) {
  final resolved = <SdkworkAuthLoginMethod>[];
  final supportsLocalCredentials = _readBool(authConfig, 'supportsLocalCredentials') ?? true;
  final supportsSessionExchange = _readBool(authConfig, 'supportsSessionExchange') ?? false;

  if (supportsLocalCredentials) {
    resolved.add(SdkworkAuthLoginMethod.password);
  }
  if (supportsSessionExchange) {
    resolved.add(SdkworkAuthLoginMethod.sessionBridge);
  }
  if (verificationPolicy.emailCodeLoginEnabled ?? false) {
    resolved.add(SdkworkAuthLoginMethod.emailCode);
  }
  if (verificationPolicy.phoneCodeLoginEnabled ?? false) {
    resolved.add(SdkworkAuthLoginMethod.phoneCode);
  }

  return resolved.isEmpty
      ? const [SdkworkAuthLoginMethod.password]
      : List<SdkworkAuthLoginMethod>.unmodifiable(resolved);
}

SdkworkAuthVerificationPolicyConfig _resolveMetadataVerificationPolicy(
  Map<String, dynamic>? authConfig,
) {
  final nestedPolicy = _readMap(authConfig, 'verificationPolicy');
  return SdkworkAuthVerificationPolicyConfig(
    emailCodeLoginEnabled: _readBool(nestedPolicy, 'emailCodeLoginEnabled')
        ?? _readBool(authConfig, 'emailCodeLoginEnabled'),
    emailRegistrationVerificationRequired:
        _readBool(nestedPolicy, 'emailRegistrationVerificationRequired')
            ?? _readBool(authConfig, 'emailRegistrationVerificationRequired'),
    phoneCodeLoginEnabled: _readBool(nestedPolicy, 'phoneCodeLoginEnabled')
        ?? _readBool(authConfig, 'phoneCodeLoginEnabled'),
    phoneRegistrationVerificationRequired:
        _readBool(nestedPolicy, 'phoneRegistrationVerificationRequired')
            ?? _readBool(authConfig, 'phoneRegistrationVerificationRequired'),
    oauthLoginEnabled: _readBool(nestedPolicy, 'oauthLoginEnabled')
        ?? _readBool(authConfig, 'oauthLoginEnabled'),
  );
}

bool _hasVerificationPolicyValues(SdkworkAuthVerificationPolicyConfig policy) {
  return policy.emailCodeLoginEnabled != null
      || policy.emailRegistrationVerificationRequired != null
      || policy.phoneCodeLoginEnabled != null
      || policy.phoneRegistrationVerificationRequired != null
      || policy.oauthLoginEnabled != null;
}

Map<String, dynamic>? _readMap(Map<String, dynamic>? source, String key) {
  final value = source?[key];
  if (value is Map<String, dynamic>) {
    return value;
  }
  if (value is Map) {
    return Map<String, dynamic>.from(value);
  }
  return null;
}

bool? _readBool(Map<String, dynamic>? source, String key) {
  final value = source?[key];
  return value is bool ? value : null;
}

List<String> _readStringList(Map<String, dynamic>? source, String key) {
  final value = source?[key];
  if (value is! List) {
    return const [];
  }
  return value.whereType<String>().toList(growable: false);
}

File: apps/sdkwork-iam-common/packages/sdkwork-iam-contracts/src/auth-runtime-metadata.ts
export type SdkworkAuthLoginMethod = "emailCode" | "password" | "phoneCode" | "sessionBridge";
export type SdkworkAuthRegisterMethod = "email" | "phone";
export type SdkworkAuthRecoveryMethod = "email" | "phone";
export type SdkworkAuthOAuthProviderRegion = "mainland" | "overseas";

export interface SdkworkAuthVerificationPolicyConfig {
  emailCodeLoginEnabled?: boolean;
  emailRegistrationVerificationRequired?: boolean;
  phoneCodeLoginEnabled?: boolean;
  phoneRegistrationVerificationRequired?: boolean;
  oauthLoginEnabled?: boolean;
}

export interface SdkworkAuthRuntimeConfig {
  loginMethods?: SdkworkAuthLoginMethod[];
  oauthLoginEnabled?: boolean;
  oauthProviderRegion?: SdkworkAuthOAuthProviderRegion;
  oauthProviders?: string[];
  sdkworkOAuthProviderEnabled?: boolean;
  qrLoginEnabled?: boolean;
  recoveryMethods?: SdkworkAuthRecoveryMethod[];
  registerMethods?: SdkworkAuthRegisterMethod[];
  verificationPolicy?: SdkworkAuthVerificationPolicyConfig;
}

export interface SdkworkCanonicalAuthMetadataLike {
  emailCodeLoginEnabled?: boolean;
  emailRegistrationVerificationRequired?: boolean;
  loginMethods?: readonly (string | undefined)[];
  oauthLoginEnabled?: boolean;
  oauthProviderRegion?: SdkworkAuthOAuthProviderRegion;
  oauthProviders?: readonly string[];
  phoneCodeLoginEnabled?: boolean;
  phoneRegistrationVerificationRequired?: boolean;
  providerKey?: string;
  qrLoginEnabled?: boolean;
  recoveryMethods?: readonly (string | undefined)[];
  registerMethods?: readonly (string | undefined)[];
  supportsLocalCredentials?: boolean;
  supportsSessionExchange?: boolean;
  sdkworkOAuthProviderEnabled?: boolean;
  verificationPolicy?: SdkworkAuthVerificationPolicyConfig;
}

export function isSdkworkAuthLoginMethod(value: string | undefined): value is SdkworkAuthLoginMethod {
  return value === "password"
    || value === "phoneCode"
    || value === "emailCode"
    || value === "sessionBridge";
}

export function isSdkworkAuthRegisterMethod(value: string | undefined): value is SdkworkAuthRegisterMethod {
  return value === "email" || value === "phone";
}

export function isSdkworkAuthRecoveryMethod(value: string | undefined): value is SdkworkAuthRecoveryMethod {
  return value === "email" || value === "phone";
}

export function isSdkworkAuthOAuthProviderRegion(
  value: string | undefined,
): value is SdkworkAuthOAuthProviderRegion {
  return value === "mainland" || value === "overseas";
}

export function resolveSdkworkAuthRuntimeConfigFromMetadata<
  TMetadata extends SdkworkCanonicalAuthMetadataLike,
>(
  authConfig?: TMetadata | null,
): SdkworkAuthRuntimeConfig {
  const verificationPolicy = resolveMetadataVerificationPolicy(authConfig);
  const loginMethods: NonNullable<SdkworkAuthRuntimeConfig["loginMethods"]> =
    authConfig?.loginMethods?.length
      ? authConfig.loginMethods
          .filter(isSdkworkAuthLoginMethod)
          .filter((method) => {
            if (method === "emailCode" && verificationPolicy.emailCodeLoginEnabled === false) {
              return false;
            }

            if (method === "phoneCode" && verificationPolicy.phoneCodeLoginEnabled === false) {
              return false;
            }

            return true;
          })
      : (() => {
          const resolvedLoginMethods: NonNullable<SdkworkAuthRuntimeConfig["loginMethods"]> = [];

          if (authConfig?.supportsLocalCredentials ?? true) {
            resolvedLoginMethods.push("password");
          }

          if (authConfig?.supportsSessionExchange) {
            resolvedLoginMethods.push("sessionBridge");
          }

          if (verificationPolicy.emailCodeLoginEnabled) {
            resolvedLoginMethods.push("emailCode");
          }

          if (verificationPolicy.phoneCodeLoginEnabled) {
            resolvedLoginMethods.push("phoneCode");
          }

          return resolvedLoginMethods.length > 0 ? resolvedLoginMethods : ["password"];
        })();
  const configuredRegisterMethods = authConfig?.registerMethods
    ?.filter(isSdkworkAuthRegisterMethod)
    ?? [];
  const registerMethods: NonNullable<SdkworkAuthRuntimeConfig["registerMethods"]> =
    configuredRegisterMethods.length > 0
      ? configuredRegisterMethods
      : ((authConfig?.supportsLocalCredentials ?? true) ? ["email"] : []);
  const recoveryMethods: NonNullable<SdkworkAuthRuntimeConfig["recoveryMethods"]> =
    authConfig?.recoveryMethods?.filter(isSdkworkAuthRecoveryMethod) ?? [];

  return {
    loginMethods,
    oauthLoginEnabled: authConfig?.oauthLoginEnabled ?? false,
    ...(authConfig?.oauthProviderRegion
      ? { oauthProviderRegion: authConfig.oauthProviderRegion }
      : {}),
    oauthProviders: authConfig?.oauthProviders ? [...authConfig.oauthProviders] : [],
    ...(typeof authConfig?.sdkworkOAuthProviderEnabled === "boolean"
      ? { sdkworkOAuthProviderEnabled: authConfig.sdkworkOAuthProviderEnabled }
      : {}),
    ...(typeof authConfig?.qrLoginEnabled === "boolean"
      ? { qrLoginEnabled: authConfig.qrLoginEnabled }
      : {}),
    recoveryMethods,
    registerMethods,
    ...(Object.keys(verificationPolicy).length > 0 ? { verificationPolicy } : {}),
  };
}

function resolveMetadataVerificationPolicy<TMetadata extends SdkworkCanonicalAuthMetadataLike>(
  authConfig?: TMetadata | null,
): SdkworkAuthVerificationPolicyConfig {
  return {
    ...(typeof authConfig?.verificationPolicy?.emailCodeLoginEnabled === "boolean"
      ? { emailCodeLoginEnabled: authConfig.verificationPolicy.emailCodeLoginEnabled }
      : typeof authConfig?.emailCodeLoginEnabled === "boolean"
        ? { emailCodeLoginEnabled: authConfig.emailCodeLoginEnabled }
        : {}),
    ...(typeof authConfig?.verificationPolicy?.emailRegistrationVerificationRequired === "boolean"
      ? { emailRegistrationVerificationRequired: authConfig.verificationPolicy.emailRegistrationVerificationRequired }
      : typeof authConfig?.emailRegistrationVerificationRequired === "boolean"
        ? { emailRegistrationVerificationRequired: authConfig.emailRegistrationVerificationRequired }
        : {}),
    ...(typeof authConfig?.verificationPolicy?.phoneCodeLoginEnabled === "boolean"
      ? { phoneCodeLoginEnabled: authConfig.verificationPolicy.phoneCodeLoginEnabled }
      : typeof authConfig?.phoneCodeLoginEnabled === "boolean"
        ? { phoneCodeLoginEnabled: authConfig.phoneCodeLoginEnabled }
        : {}),
    ...(typeof authConfig?.verificationPolicy?.phoneRegistrationVerificationRequired === "boolean"
      ? { phoneRegistrationVerificationRequired: authConfig.verificationPolicy.phoneRegistrationVerificationRequired }
      : typeof authConfig?.phoneRegistrationVerificationRequired === "boolean"
        ? { phoneRegistrationVerificationRequired: authConfig.phoneRegistrationVerificationRequired }
        : {}),
  };
}
